lib/base.js

```diff
@@ -0,0 +1,790 @@
+  this.meta = new Meta(meta || {});
+  this.attributes = attributes || {};
+  this.content = content || null;
+  this._attributeElementKeys = [];
+};
+
+_.extend(ElementType.prototype, {
+  primitive: function() {
+    return;
+  },
+
+  toValue: function() {
+    return this.content;
+  },
+
+  toRefract: function(options) {
+    var attributes = this.convertAttributesToRefract('toRefract');
+    var initial = {
+      element: this.element,
+      meta: this.meta.toObject(),
+      attributes: attributes,
+      content: this.content
+    };
+    return _.extend(initial, options || {});
+  },
+
+  toCompactRefract: function() {
+    var attributes = this.convertAttributesToRefract('toCompactRefract');
+    return [this.element, this.meta.toObject(), attributes, this.content];
+  },
+
+  /*
+   * Some attributes may be elements. This is domain-specific knowledge, so
+   * a subclass *MUST* define the attribute element names to convert. This
+   * method handles the actual serialization to refract.
+   */
+  convertAttributesToRefract: function(functionName) {
+    var attributes = {};
+    var keys = _.keys(this.attributes);
+
+    for (var i = 0; i < keys.length; i++) {
+      var key = keys[i];
+      if (this._attributeElementKeys.indexOf(key) !== -1) {
+        attributes[key] = this.attributes[key][functionName]();
+      } else {
+        attributes[key] = this.attributes[key];
+      }
+    }
+
+    return attributes;
+  },
+
+  /*
+   * Some attributes may be elements. This is domain-specific knowledge, so
+   * a subclass *MUST* define the attribute element names to convert. This
+   * method handles the actual conversion when loading.
+   */
+  convertAttributesToElements: function(conversionFunc) {
+    for (var i = 0; i < this._attributeElementKeys.length; i++) {
+      var key = this._attributeElementKeys[i];
+
+      if (this.attributes[key]) {
+        this.attributes[key] = conversionFunc(this.attributes[key]);
+      }
+
+    }
+  },
+
+  fromRefract: function(dom) {
+    this.meta = new Meta(dom.meta);
+    this.attributes = dom.attributes;
+    this.content = dom.content;
+
+    this.convertAttributesToElements(function(attribute) {
+      return registry.fromRefract(attribute);
+    });
+
+    if (this.element !== dom.element) {
+      this.element = dom.element;
+    }
+
+    return this;
+  },
+
+  fromCompactRefract: function(tuple) {
+    this.meta = new Meta(tuple[1]);
+    this.attributes = tuple[2];
+    this.content = tuple[3];
+
+    this.convertAttributesToElements(function(attribute) {
+      registry.fromCompactRefract(attribute);
+    });
+
+    if (this.element !== tuple[0]) {
+      this.element = tuple[0];
+    }
+
+    return this;
+  },
+
+  set: function(content) {
+    this.content = content;
+    return this;
+  },
+
+  equals: function(value) {
+    return _.isEqual(this.toValue(), value);
+  }
+});
+
+Object.defineProperty(ElementType.prototype, 'element', {
```

Comment:
Is there any way to do this within the `extend` syntax above? This feels a little out of place, but it could work okay.